src/Transport/Receiving/MessagePump.cs

```diff
@@ -296,36 +298,38 @@ await receiver.SafeCompleteAsync(transportSettings.TransportTransactionMode, loc
                     {
                         var transportTransaction = CreateTransportTransaction(message.PartitionKey, transaction);
 
-                        var errorContext = new ErrorContext(ex, message.GetNServiceBusHeaders(), messageId, body, transportTransaction, message.SystemProperties.DeliveryCount, contextBag);
+                        var errorContext = new ErrorContext(ex, message.GetNServiceBusHeaders(), messageId, body, transportTransaction, message.DeliveryCount, contextBag);
 
                         result = await onError(errorContext, messageProcessingCancellationToken).ConfigureAwait(false);
 
                         if (result == ErrorHandleResult.Handled)
                         {
-                            await receiver.SafeCompleteAsync(transportSettings.TransportTransactionMode, lockToken, transaction, messageProcessingCancellationToken).ConfigureAwait(false);
+                            await receiver.SafeCompleteMessageAsync(message, transportSettings.TransportTransactionMode, transaction, cancellationToken: messageProcessingCancellationToken).ConfigureAwait(false);
                         }
 
                         transaction?.Commit();
                     }
 
                     if (result == ErrorHandleResult.RetryRequired)
                     {
-                        await receiver.SafeAbandonAsync(transportSettings.TransportTransactionMode, lockToken, cancellationToken: messageProcessingCancellationToken).ConfigureAwait(false);
+                        await receiver.SafeAbandonMessageAsync(message, transportSettings.TransportTransactionMode, cancellationToken: messageProcessingCancellationToken).ConfigureAwait(false);
                     }
                 }
-                catch (Exception onErrorEx) when (onErrorEx.IsCausedBy(messageProcessingCancellationToken))
+                catch (ServiceBusException onErrorEx) when (onErrorEx.Reason == ServiceBusFailureReason.MessageLockLost || onErrorEx.Reason == ServiceBusFailureReason.ServiceTimeout)
                 {
-                    throw;
+                    Logger.Debug("Failed to execute recoverability.", onErrorEx);
+
+                    await receiver.AbandonMessageAsync(message, cancellationToken: messageProcessingCancellationToken).ConfigureAwait(false);
```

Comment:
@andreasohlund @WojcikMike do you remember why this line doesn't use `SafeAbandonAsync` like the other places?

Reply:
If I remember correctly there is no need for SafeXyz when you don't want to be "atomic with the receive operation" and since this happens when recoverability has failed you would always want the message to go back to the queue no matter what. (@WojcikMike @mauroservienti please confirm)

That said I don't think it would hurt to use the SafeX version here as well for consistency?

Reply:
I honestly do not remember, sorry.